refactor: wallet, remove global 'ArgsManager' dependency

furszy opened the pull request:
Structurally, the wallet class shouldn't access the global `ArgsManager` class, its internal behavior shouldn't be coupled to a global command line args parsing object.

So this PR migrates the only two places where we depend on it: (1) the keypool size, and (2) the "-walletnotify" script. And cleans up the, now unneeded, wallet `ArgsManager` ref member.

Extra note:
In the process of removing the args ref member, discovered and fixed files that were invalidly depending on the wallet header including `util/system.h`.

furszy committed:
wallet: set keypool_size instead of access global args manager

TheCharlatan commented on src/wallet/wallet.cpp:
```diff
@@ -2883,6 +2883,9 @@ std::shared_ptr<CWallet> CWallet::Create(WalletContext& context, const std::stri
     // TODO: Can't use std::make_shared because we need a custom deleter but
     // should be possible to use std::allocate_shared.
     std::shared_ptr<CWallet> walletInstance(new CWallet(chain, name, args, std::move(database)), ReleaseWallet);
+    walletInstance->m_keypool_size = std::max(args.GetIntArg("-keypool", DEFAULT_KEYPOOL_SIZE), (int64_t) 1);
```
There might be a good reason for this, but why not initialize this member field in the header?

furszy replied:
where in the header exactly? constructor?

TheCharlatan replied:
Yes:
```
         : m_args(args),
           m_chain(chain),
           m_name(name),
-          m_database(std::move(database))
+          m_database(std::move(database)),
+          m_keypool_size(std::max(args.GetIntArg("-keypool", DEFAULT_KEYPOOL_SIZE), (int64_t) 1))
```

furszy replied:
Structurally speaking, the wallet class shouldn't have access to the global `ArgsManager` class, its internal behavior shouldn't be coupled to a global command line args parsing object. So setting this on the constructor would just hide the non-desire coupling.

Could actually go further here and totally remove the `m_args` member from the wallet now. Just need to keep track of the "-walletnotify" script.
Let me see how big this change is.

furszy replied:
Updated the PR with this changes.

TheCharlatan replied:
Great, that seems conceptually sound to me!

sedited reviewed (commented):
Concept ACK

furszy committed:
wallet: set '-walletnotify' script instead of access global args manager

furszy committed:
refactor: wallet, remove global 'ArgsManager' access

we are not using it anymore

furszy committed:
refactor: remove <util/system.h> include from wallet.h

Since we no longer store a ref to the global `ArgsManager`
inside the wallet, we can move the util/system.h
include to the cpp.

This dependency removal opened a can of worms, as few
other places were, invalidly, depending on the wallet's
header including it.

fanquake commented:
Concept ACK

furszy commented:
rebased, conflicts solved.

pablomartin4btc reviewed (commented):
code review ACK, pending of testing it.

Instead of removing it all in once you could extract the gui bit/ qt part (6/26 ~ 20%) into another PR in `/bitcoin-core/gui`, so you reduce the amount of files modified touched by the PR - perhaps other splits can be performed as well.

furszy commented:
> Instead of removing it all in once you could extract the gui bit/ qt part (6/26 ~ 20%) into another PR in /bitcoin-core/gui, so you reduce the amount of files modified touched by the PR - perhaps other splits can be performed as well.

There are two commits that slightly touch the GUI.

6b578a37 because the GUI tests create a `CWallet` whose constructor signature has changed (due the removal of the global args object ref).
In this scenario, nothing can be done, the GUI test must be changed within the same commit.

Then 9022473, which modifications are originated from the no longer needed `util/system.h` include removed from the `wallet.h` file.
In this scenario, moving the changes into a GUI-only PR would mean that the GUI PR would touch the backend code, which usually is a no-go. Plus, that commit alone, based on the subset of changes that contains, it's natural to do it here as it doesn't worth to push it in a standalone follow-up.

pablomartin4btc commented:
> 
> In this situation, separating the changes into a separate GUI-only PR would mean the GUI PR would impact the backend code, which is typically not recommended. Furthermore, given the limited scope of the changes in that specific commit, it makes more sense to keep it in this PR rather than creating a separate follow-up.

I see, yeah, I understand now it would take some time to split the PRs and as you said the scope is very specific, but for future reference, perhaps you can consider this, as some reviewers could take the amount of files being touched into account. I like this refactoring in general, as I said, I'll test it asap. Thanks!

pablomartin4btc reviewed (approved):
tested ACK 9022473ff167f2677ccad27d98705c3887bf9f96 (unit tests, functionals extended & qt tests)

hebasto commented:
Concept ACK.

hebasto reviewed (commented):
Approach ACK 9022473ff167f2677ccad27d98705c3887bf9f96, I've reviewed code and verified changes in the included headers with the IWYU tool.

A few style nits have been suggested by [`clang-format-diff.py`](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy).

furszy commented on src/wallet/wallet.h:
```diff
@@ -642,6 +638,12 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     CAmount m_default_max_tx_fee{DEFAULT_TRANSACTION_MAXFEE};
 
+    /** Number of pre-generated keys/scripts by each spkm (part of the look-ahead process, used to detect payments) */
+    uint64_t m_keypool_size{DEFAULT_KEYPOOL_SIZE};
```
Why only for bit-wise operations? It's just a standard data type that represents an unsigned 64-bit integer.
Didn't have any specific reason, just that is more concise and specific. Could change it to a bare unsigned int too.

hebasto replied:
See:
- https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#es101-use-unsigned-types-for-bit-manipulation
- https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#es102-use-signed-types-for-arithmetic
- [Signed and Unsigned Types in Interfaces](https://www.aristeia.com/Papers/C++ReportColumns/sep95.pdf)

Suggesting:
```suggestion
    int64_t m_keypool_size{DEFAULT_KEYPOOL_SIZE};
```

furszy replied:
ah, you are suggesting to change it to a signed type. Sure.
Pushed.

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -2913,7 +2914,11 @@ std::shared_ptr<CWallet> CWallet::Create(WalletContext& context, const std::stri
     // should be possible to use std::allocate_shared.
-    std::shared_ptr<CWallet> walletInstance(new CWallet(chain, name, args, std::move(database)), ReleaseWallet);
+    std::shared_ptr<CWallet> walletInstance(new CWallet(chain, name, std::move(database)), ReleaseWallet);
+    walletInstance->m_keypool_size = std::max(args.GetIntArg("-keypool", DEFAULT_KEYPOOL_SIZE), (int64_t) 1);
```
pushed

furszy commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -286,6 +286,9 @@ class LegacyScriptPubKeyMan : public ScriptPubKeyMan, public FillableSigningProv
     int64_t nTimeFirstKey GUARDED_BY(cs_KeyStore) = 0;
 
+    //! Number of pre-generated keys/scripts (part of the look-ahead process, used to detect payments)
+    uint64_t m_keypool_size GUARDED_BY(cs_KeyStore) {DEFAULT_KEYPOOL_SIZE};
```
pushed

TheCharlatan commented on src/wallet/rpc/backup.cpp:
```diff
@@ -1478,7 +1478,7 @@ static UniValue ProcessDescriptorImport(CWallet& wallet, const UniValue& data, c
                 warnings.push_back("Range not given, using default keypool range");
                 range_start = 0;
-                range_end = gArgs.GetIntArg("-keypool", DEFAULT_KEYPOOL_SIZE);
+                range_end = std::max(gArgs.GetIntArg("-keypool", DEFAULT_KEYPOOL_SIZE), (int64_t)1);
```
I might have missed something, but why not use the `m_keypool_size` from the available wallet here?

```
diff --git a/src/wallet/rpc/backup.cpp b/src/wallet/rpc/backup.cpp
index 64253789ec..d9c6755bcb 100644
--- a/src/wallet/rpc/backup.cpp
+++ b/src/wallet/rpc/backup.cpp
@@ -15,7 +15,6 @@
 #include <script/standard.h>
 #include <sync.h>
 #include <util/bip32.h>
-#include <util/system.h>
 #include <util/time.h>
 #include <util/translation.h>
 #include <wallet/rpc/util.h>
@@ -1480,7 +1479,7 @@ static UniValue ProcessDescriptorImport(CWallet& wallet, const UniValue& data, c
             } else {
                 warnings.push_back("Range not given, using default keypool range");
                 range_start = 0;
-                range_end = gArgs.GetIntArg("-keypool", DEFAULT_KEYPOOL_SIZE);
+                range_end = wallet.m_keypool_size;
             }
             next_index = range_start;
```

furszy replied:
ha, because I'm still sleepy, good catch. Pushing it..

furszy commented on src/wallet/rpc/backup.cpp:
```diff
@@ -1478,7 +1478,7 @@ static UniValue ProcessDescriptorImport(CWallet& wallet, const UniValue& data, c
                 warnings.push_back("Range not given, using default keypool range");
                 range_start = 0;
-                range_end = gArgs.GetIntArg("-keypool", DEFAULT_KEYPOOL_SIZE);
+                range_end = wallet.m_keypool_size;
```
done

furszy commented:
Updated per feedback. Thanks. Tiny [diff](https://github.com/bitcoin/bitcoin/compare/ebb14064a0589bc56b71e267204696accc577036..fafaa7a7e15a49da0214ce9546a42117c63e611f)

sedited reviewed (approved):
Code review ACK fafaa7a7e15a49da0214ce9546a42117c63e611f

I would take this one step further though in a follow-up PR and replace the `ArgsManager` in the `WalletContext` with an options struct that gets passed to each `CWallet` instance.

hebasto commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -555,6 +560,9 @@ class DescriptorScriptPubKeyMan : public ScriptPubKeyMan
     bool m_decryption_thoroughly_checked = false;
 
+    //! Number of pre-generated keys/scripts (part of the look-ahead process, used to detect payments)
+    uint64_t m_keypool_size GUARDED_BY(cs_desc_man){DEFAULT_KEYPOOL_SIZE};
```
Sorry for not being explicit in my previous comments. There are more cases where `uint64_t` could be replaced with `int64_t`:
```suggestion
    int64_t m_keypool_size GUARDED_BY(cs_desc_man){DEFAULT_KEYPOOL_SIZE};
```

and parameter `keypool_size` in (member) functions.

furszy replied:
np, small miscommunication. Done now, all of them tackled now. Thanks!

furszy commented:
Feedback tackled.
Moved remaining variables to signed type. Small [diff](https://github.com/bitcoin/bitcoin/compare/fafaa7a7e15a49da0214ce9546a42117c63e611f..d90b54b4aa18d60faa0075348fbb29a513558098).

hebasto commented on src/wallet/wallet.cpp:
```diff
@@ -2913,7 +2914,11 @@ std::shared_ptr<CWallet> CWallet::Create(WalletContext& context, const std::stri
     // should be possible to use std::allocate_shared.
-    std::shared_ptr<CWallet> walletInstance(new CWallet(chain, name, args, std::move(database)), ReleaseWallet);
+    std::shared_ptr<CWallet> walletInstance(new CWallet(chain, name, std::move(database)), ReleaseWallet);
+    walletInstance->m_keypool_size = std::max(args.GetIntArg("-keypool", DEFAULT_KEYPOOL_SIZE), (int64_t)1);
```
nanonit: slightly prefer
```suggestion
    walletInstance->m_keypool_size = std::max(args.GetIntArg("-keypool", DEFAULT_KEYPOOL_SIZE), int64_t{1});
```

as it was before, and is used recently in our code base.

furszy replied:
done

hebasto reviewed (approved):
ACK d90b54b4aa18d60faa0075348fbb29a513558098

hebasto reviewed (approved):
re-ACK fa1ce3dd62be739b579373287c5501d99fcf9c17

sedited reviewed (approved):
re-ACK fa1ce3dd62be739b579373287c5501d99fcf9c17

fanquake commented:
@achow101 can you take a look here

achow101 commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -365,6 +368,8 @@ class LegacyScriptPubKeyMan : public ScriptPubKeyMan, public FillableSigningProv
 public:
     using ScriptPubKeyMan::ScriptPubKeyMan;
 
+    LegacyScriptPubKeyMan(WalletStorage& storage, int64_t keypool_size) : ScriptPubKeyMan(storage), m_keypool_size(keypool_size) {}
```
In e86214692d3fe7a8c835e07f7ebd47a9afb9fec9 "wallet: set keypool_size instead of access global args manager"

The `using ScriptPubKeyMan::ScriptPubKeyMan;` line above should be removed so that this new constructor is the only constructor to avoid possible programming errors where the keypool size is not passed in.

furszy replied:
done

furszy commented:
Updated per feedback. Thanks.
Single line removal [diff](https://github.com/bitcoin/bitcoin/compare/fa1ce3dd62be739b579373287c5501d99fcf9c17..52f4d567d69425dfd514489079db80483024a80d).

sedited reviewed (approved):
Re-ACK 52f4d567d69425dfd514489079db80483024a80d

hebasto reviewed (approved):
re-ACK 52f4d567d69425dfd514489079db80483024a80d

achow101 commented:
ACK 52f4d567d69425dfd514489079db80483024a80d

pablomartin4btc reviewed (commented):
re-ACK 52f4d567d69425dfd514489079db80483024a80d; checked all updates, good stuff!

ponury1990 reviewed (commented):
Witam chciałem spytać o ewentualne inne metody odebrania profitu nie obaw co do moich umiejętności tworzenia prawdę pisząc to ja nie za wiele wiem no a zdążyłem zebrać masę krytyki tylko za co ? Gratuluje wszystkim co za pierwszym razem gdy mieli styczność z projektem byli mądrzejsi niż Wikipedia przykre ze tak prestiżowy i profesjonalny portal jakim jest Git chowa młodych do jednego wora nie pytając o przyczynę. P.S strach sie zapytać o cokolwiek

pablomartin4btc commented:
Found an [issue](https://github.com/bitcoin/bitcoin/issues/21005) where @MarcoFalke mentioned the need of removing gArgs from dfferent places and @ryanofsky's [comment](https://github.com/bitcoin-core/gui/pull/75#issuecomment-689675368) regarding the GUI specifically, which seem very interesting so I wanted to add the links here.
